src/arch/src/aarch64/regs.rs

```diff
@@ -209,6 +229,74 @@ pub fn read_mpidr(vcpu: &VcpuFd) -> Result<u64> {
     vcpu.get_one_reg(MPIDR_EL1).map_err(Error::GetSysRegister)
 }
 
+/// Get the state of the system and core registers.
+///
+/// # Arguments
+///
+/// * `vcpu` - Structure for the VCPU that holds the VCPU's fd.
+/// * `state` - Structure for returning the state of the system registers.
+pub fn save_vcpu_registers(vcpu: &VcpuFd, state: &mut Vec<kvm_one_reg>) -> Result<()> {
+    // Call KVM_GET_REG_LIST to get all registers available to the guest. For ArmV8 there are
+    // around 500 registers.
+    let mut reg_list = RegList::new(512);
+    vcpu.get_reg_list(&mut reg_list)
+        .map_err(Error::GetRegList)?;
+
+    // At this point reg_list should contain: core registers and system registers.
+    // The register list contains the number of registers and their ids. We will be needing to
+    // call KVM_GET_ONE_REG on each id in order to save all of them. We carve out from the list
+    // the core registers which are represented in the kernel by kvm_regs structure and for which
+    // we can calculate the id based on the offset in the structure.
+    reg_list.retain(|regid| is_core_register(*regid) || is_system_register(*regid));
+
+    // Now, for the rest of the registers left in the previously fetched register list, we are
+    // simply calling KVM_GET_ONE_REG.
+    let indices = reg_list.as_slice();
+    for index in indices.iter() {
```

Comment:
It might be more efficient to just do something like `reg_list.as_slice().iter().filter(|regid| is_core_register(*regid) || is_system_register(*regid))` ?  `retain` removes elements from the `FamStruct`. `filter` should only do the check.

Reply:
I will try it. Did not know there exists 'filter'.